### PR TITLE
Final integrations & finishing up work

### PR DESCRIPTION
**Pull request by pjohnmeyer**
Depends on and is based on #5 

@kramasamy This pull is a work in progress and represents the work being done on:

- [x] [Logging](https://github.com/streamlio/fluent-bit/projects/1#card-17402086)
- [x] [Robust message production](https://github.com/streamlio/fluent-bit/projects/1#card-17402095)
- [x] [Valgrind testing + fixes](https://github.com/streamlio/fluent-bit/projects/1#card-17532892)

## Current Valgrind summary
Key takeway: we are leaking some memory, but it doesn't appear to be increasing based on the duration of the run, so it's must all be allocations at startup that aren't deallocated. Not sure how much effort to put into that -- the OS is really good at freeing up memory.

### Baseline `fluent-bit -i cpu -o null`
```
==12745== HEAP SUMMARY:
==12745==     in use at exit: 3,176 bytes in 47 blocks
==12745==   total heap usage: 4,780 allocs, 4,733 frees, 13,713,784 bytes allocated
==12745== 
==12745== LEAK SUMMARY:
==12745==    definitely lost: 0 bytes in 0 blocks
==12745==    indirectly lost: 0 bytes in 0 blocks
==12745==      possibly lost: 0 bytes in 0 blocks
==12745==    still reachable: 3,176 bytes in 47 blocks
==12745==         suppressed: 0 bytes in 0 blocks
==12745== Reachable blocks (those to which a pointer was found) are not shown.
==12745== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==12745== 
==12745== For counts of detected and suppressed errors, rerun with: -v
==12745== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

### Pulsar run 1 `fluent-bit -i cpu -o pulsar`
```
==11449== HEAP SUMMARY:
==11449==     in use at exit: 355,108 bytes in 155 blocks
==11449==   total heap usage: 11,707 allocs, 11,552 frees, 11,575,582 bytes allocated
==11449==
==11449== LEAK SUMMARY:
==11449==    definitely lost: 108 bytes in 1 blocks
==11449==    indirectly lost: 0 bytes in 0 blocks
==11449==      possibly lost: 137,004 bytes in 32 blocks
==11449==    still reachable: 217,996 bytes in 122 blocks
==11449==         suppressed: 0 bytes in 0 blocks
==11449== Reachable blocks (those to which a pointer was found) are not shown.
==11449== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==11449== 
==11449== For counts of detected and suppressed errors, rerun with: -v
==11449== ERROR SUMMARY: 33 errors from 33 contexts (suppressed: 0 from 0)
```

### Pulsar run 2 (run ~twice as long):
```
==11968== HEAP SUMMARY:
==11968==     in use at exit: 355,108 bytes in 155 blocks
==11968==   total heap usage: 15,751 allocs, 15,596 frees, 24,444,630 bytes allocated
==11968== 
==11968== LEAK SUMMARY:
==11968==    definitely lost: 108 bytes in 1 blocks
==11968==    indirectly lost: 0 bytes in 0 blocks
==11968==      possibly lost: 137,004 bytes in 32 blocks
==11968==    still reachable: 217,996 bytes in 122 blocks
==11968==         suppressed: 0 bytes in 0 blocks
==11968== Reachable blocks (those to which a pointer was found) are not shown.
==11968== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==11968== 
==11968== For counts of detected and suppressed errors, rerun with: -v
==11968== ERROR SUMMARY: 33 errors from 33 contexts (suppressed: 0 from 0)
```

### CODE REVIEW

**Comment by kramasamy** [comment]
@pjohnmeyer -ok. do we know if the memory leak is in pulsar or fluent bit?

**Comment by pjohnmeyer** [comment]
@kramasamy I'm going to do a more scientific test with a fixed input set to be sure. I ran again with `-i cpu -o pulsar` for about a day and a half while I was out of town, and the two leaks with traces in the output appear to originate in pulsar client code.

**Comment by pjohnmeyer** [comment]
Some updates on the valgrind results. Attached is a full output file from:

1. Invoking CMake on fluent-bit with `-DFLB_OUT_PULSAR=On -DFLB_SHARED_LIB=Off -DFLB_VALGRIND=On -DFLB_EXAMPLES=Off`
2. Building `fluent-bit-bin`
3. Running `gen-input.sh | valgrind --leak-check=full --track-origins=yes ./bin/fluent-bit -i stdin -o pulsar -p Topic=non-persistent://public/default/fluent-bit`

Where `gen-input.sh` looks like this:

```bash
#!/usr/bin/env bash
for run in {1..10000}
do
  echo -n "{\"key\": \"$run\"}"
  sleep 0.001
done
```

First, there are two instances of valgrind finding "Conditional jump or move depends on uninitialised value(s)", both of which occur somewhere in the stack for the stdin input plugin. We can discard those.

The three leaks that show up appear to originate in boost and pulsar code. My code does appear in the stack of at least one, but the allocations originate in the pulsar libraries. My suspicion is these are startup/shutdown-related leaks, as I was also able to run this for a day and a half with no visible memory growth.

[valgrind.log](https://github.com/streamlio/fluent-bit/files/2872604/valgrind.log)

In short, I don't think there is any more work to do on memory management.

**Comment by kramasamy** [comment]
Fixed some leaks in our end - https://github.com/apache/pulsar/pull/3618

**Comment by david-streamlio** [comment]
@pjohnmeyer Any idea why the build is failing now?

**Comment by pjohnmeyer** [comment]
@david-streamlio The failing test is an input plugin test:

`3/30 Test  #3: flb-rt-in_head ...................***Failed    2.98 sec`

Some of the specs are erratic and prone to intermittent failures.

File: tests/runtime/helpers/out_pulsar/build_message.cc [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include "build_message.h"
+
+#include <pulsar/Message.h>
+#include <pulsar/MessageBuilder.h>
+#include "../../../../plugins/out_pulsar/pulsar-client-2.2.1/pulsar-client-cpp/lib/c/c_structs.h"
```
Comment (merlimat) [inline tests/runtime/helpers/out_pulsar/build_message.cc]: Is there an easy way to avoid the version specific include? 
Why does the message need to be built in this case? Should the "build" be exposed in the C API ?

Comment (pjohnmeyer) [reply tests/runtime/helpers/out_pulsar/build_message.cc]: > Is there an easy way to avoid the version specific include?

There are a few ways, but I'm a little hesitant... this is the kind of thing I like to call a "good nuisance." We want this to be noisy and annoying because it's "bad" that we're including internal lib files, and so we want it to be revisited if the API changes in a way that it is no longer necessary.

> Why does the message need to be built in this case?

The message needs to be built so that in unit tests we can call the `pulsar_message_get_` functions.

> Should the "build" be exposed in the C API ?

That's a good question. From a pure API design perspective I would say "no" -- you shouldn't expose anything you do not intend to be used. But, as a user of your API, I find it useful for testing my code. As a compromise, I would suggest putting it in a separate header, or behind a pre-compiler guard, so I have to thoughtfully choose to make that function accessible to my client code.

Comment (pjohnmeyer) [reply tests/runtime/helpers/out_pulsar/build_message.cc]: @merlimat This diff is one way we could address the version-specific include issue. Let me know if you want me to push it:

```diff
diff --git a/plugins/out_pulsar/CMakeLists.txt b/plugins/out_pulsar/CMakeLists.txt
index 57789643..a82dc05d 100644
--- a/plugins/out_pulsar/CMakeLists.txt
+++ b/plugins/out_pulsar/CMakeLists.txt
@@ -30,6 +30,9 @@ include_directories(curl-7.61.0/include/)
 
 add_subdirectory(pulsar-client-2.2.1/pulsar-client-cpp EXCLUDE_FROM_ALL)
 
+# Added because the test project needs access to a file in /lib
+target_include_directories(pulsarStatic PUBLIC pulsar-client-2.2.1)
+
 # Fluent Bit Pulsar Output plugin
 set(src
   pulsar.c
diff --git a/tests/runtime/helpers/out_pulsar/build_message.cc b/tests/runtime/helpers/out_pulsar/build_message.cc
index 28fb97e5..430553e5 100644
--- a/tests/runtime/helpers/out_pulsar/build_message.cc
+++ b/tests/runtime/helpers/out_pulsar/build_message.cc
@@ -22,7 +22,7 @@
 
 #include <pulsar/Message.h>
 #include <pulsar/MessageBuilder.h>
-#include "../../../../plugins/out_pulsar/pulsar-client-2.2.1/pulsar-client-cpp/lib/c/c_structs.h"
+#include <pulsar-client-cpp/lib/c/c_structs.h>
 
 void pulsar_build_message(pulsar_message_t * message)
 {
```
